request.go

```diff
@@ -1,196 +1,126 @@
 package rpc
 
 import (
-	"fmt"
-
 	"golang.org/x/net/context"
 )
 
 type request interface {
-	Message() *message
+	rpcMessage
 	CancelFunc() context.CancelFunc
-	Reply(encoder, LogInterface) error
-	Serve(byteReadingDecoder, encoder, *ServeHandlerDescription, WrapErrorFunc, LogInterface)
-	LogInvocation(log LogInterface, err error, arg interface{})
-	LogCompletion(log LogInterface, err error)
+	Reply(encoder, interface{}, interface{}) error
+	Serve(encoder, *ServeHandlerDescription, WrapErrorFunc)
+	LogInvocation(err error)
+	LogCompletion(res interface{}, err error)
 }
 
 type requestImpl struct {
-	message
 	ctx        context.Context
 	cancelFunc context.CancelFunc
-}
-
-func (req *requestImpl) Message() *message {
-	return &req.message
+	log        LogInterface
 }
 
 func (req *requestImpl) CancelFunc() context.CancelFunc {
 	return req.cancelFunc
 }
 
-func (r *requestImpl) LogInvocation(LogInterface, error, interface{}) {}
-func (r *requestImpl) LogCompletion(LogInterface, error)              {}
-func (r *requestImpl) Reply(encoder, LogInterface) error              { return nil }
-func (r *requestImpl) Serve(byteReadingDecoder, encoder, *ServeHandlerDescription, WrapErrorFunc, LogInterface) {
-}
-
-func (req *requestImpl) getArg(receiver decoder, handler *ServeHandlerDescription) (interface{}, error) {
-	arg := handler.MakeArg()
-	err := decodeMessage(receiver, req.Message(), arg)
-	return arg, err
+func (r *requestImpl) LogInvocation(error)                           {}
```

Comment:
do you need these default implementations? would rather have the callRequest/notifyRequest fill in the ones they don't need explicitly (just Reply for notifyRequest, right?)